### ISSUE TITLE
Custom batch_size for the pipe method crashed the program

### ISSUE THREAD

**Issue by lingvisa** [explosion/spaCy#5779]
Spacy 2.3, Chinese models.

`nlpdocs = self.nlp.pipe(texts, disable=disable, n_process=-1)`

When I changed this line to:
`nlpdocs = self.nlp.pipe(texts, disable=disable, batch_size=10000, n_process=-1)`

My program crashed after it ran for a while. It was processing a text file of 110000 lines, around 60MB. The error message is:
```
Process Process-10:
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/data0/congmin/wb-nlp-tool/spaCy-weibo/spacy/language.py", line 1169, in _apply_pipes
    sender.send([doc.to_bytes() for doc in docs])
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 393, in _send_bytes
    header = struct.pack("!i", n)
struct.error: 'i' format requires -2147483648 <= number <= 2147483647
```
It doesn't always crash though. When I enable less components, it works fine with batch_size = 10000.

**Comment by lingvisa** [comment]
In another situation, I didn't change the default batch_size but still got this error:
```
Process Process-10: 
Traceback (most recent call last):
  File "/usr/lib/python3.6/multiprocessing/process.py", line 258, in _bootstrap
    self.run()
  File "/usr/lib/python3.6/multiprocessing/process.py", line 93, in run
    self._target(*self._args, **self._kwargs)
  File "/data0/congmin/wb-nlp-tool/spaCy-weibo/spacy/language.py", line 1169, in _apply_pipes
    sender.send([doc.to_bytes() for doc in docs])
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 206, in send
    self._send_bytes(_ForkingPickler.dumps(obj))
  File "/usr/lib/python3.6/multiprocessing/connection.py", line 393, in _send_bytes
    header = struct.pack("!i", n)
struct.error: 'i' format requires -2147483648 <= number <= 2147483647
```
The reason might be I started a number of my application instances. The single file isn't too big, 100MB at most. It might be due to memory issues, etc. It seems it reached the max integer for some reason. 

A general question is, for a single machine with multiple cores, if I already set the n_processor = -1, it can make use of all available cores through spacy multiprocessing.  In such a case, on the same machine to start multiple instances of my application won't speed up further. Right? By start multiple instances of my applicatoins, I mean that I might in a for loop, i execute the same command 'python my_app.py'.

**Comment by adrianeboyd** [comment]
This looks like it's an underlying constraint in certain versions of pickle used with multiprocessing in terms of how the length of the data is encoded (with a signed 32-bit integer): https://bugs.python.org/issue17560. It looks like it might be addressed in python 3.8, but I haven't tested anything.

In general, passing this much data around with multiprocessing is going to be inefficient and it'd be better to restructure how data is shared with the child processes. One possibility is using shared memory as discussed in #5051 (also restricted to python 3.8+).

Things will get slower if you have multiple processes that are each trying to use all the CPUs.
